### PR TITLE
[BCG][DSV4-Flash][CP] Support DeepSeek V4 prefill with interleave CP and breakable CUDA graphs

### PR DESCRIPTION
**Pull request by BJWang-ant**
## Summary

Enable DeepSeek V4 prefill with interleave context parallelism (CP) and breakable CUDA graphs (BCG). Previously, CP metadata and buffer layouts were not compatible with DSV4 graph capture and replay.

- Add support for interleave CP with the DSV4 attention backend under the existing PP=1 and attention CP size=TP size constraints.
- Keep per-rank replay buffers and graph write capacities stable while preserving actual request lengths for attention metadata.
- Gather DSV4 hidden states and DSpark auxiliary outputs, restore global token order, and trim padding before logits processing.
- Fall back to eager execution when a batch does not activate CP.

Also fix large CP prefill batches entering the sparse prefill path through the query-count heuristic, despite its environment switch being disabled. That path builds global query offsets against CP-local buffers, which can cause out-of-bounds writes. Route active CP batches through paged attention and validate sparse metadata buffer capacity before launching kernels.

## Validation

- Passed 8 isolated CPU regression checks covering configuration eligibility, capture buckets, replay buffer stability, output gathering, metadata lengths, and sparse prefill routing and bounds checks.
- Passed Python syntax and diff whitespace checks.
- Observed successful GPU graph capture and small-request replay before the final sparse prefill routing fix.

Full GPU end-to-end validation of the final patch is pending. CPU checks use mocked distributed collectives and do not validate CUDA kernels.

## Perf
I deployed the 1p1d service of v4-flash on B200.
chunked-prefill-size=8192
max-prefill-tokens=8192
cuda-graph-max-bs-prefill=8192
server command:
```
# prefill
GLOO_SOCKET_IFNAME=eth0 \
      NCCL_SOCKET_IFNAME=eth0 \
      SGLANG_REQUEST_STATE_WAIT_TIMEOUT=600 \
      MC_TE_METRIC=1 \
      MC_TE_METRIC_INTERVAL_SECONDS=1 \
      MC_INTRA_NVLINK=true \
      MC_INTRANODE_NVLINK=true \
      CUDA_VISIBLE_DEVICES=0,1 \
      SGLANG_ENABLE_CP_V2=1 \
      SGLANG_OPT_DEEPGEMM_MEGA_MOE_NUM_MAX_TOKENS_PER_RANK=8192 \
      SGLANG_MOONCAKE_CUSTOM_MEM_POOL=INTRA_NODE_NVLINK \
      SGLANG_DISAGGREGATION_ALL_CP_RANKS_TRANSFER=1 \
      sglang serve \
          --trust-remote-code \
          --model-path models/deepseek-ai/DeepSeek-V4-Flash-0731 \
          --tp 2 \
          --ep 2 \
          --enable-prefill-cp \
          --cp-strategy interleave \
          --nnodes 1 \
          --node-rank 0 \
          --moe-a2a-backend megamoe \
          --enable-w4a4-mxfp4-megamoe \
          --moe-runner-backend flashinfer_mxfp4 \
          --disaggregation-mode prefill \
          --disaggregation-transfer-backend mooncake \
          --speculative-algorithm DSPARK \
          --chunked-prefill-size 8192 \
          --max-prefill-tokens 8192 \
          --enable-deepseek-v4-fp4-indexer \
          --max-running-requests 256 \
          --swa-full-tokens-ratio 0.05 \
          --mem-fraction-static 0.8 \
          --detokenizer-worker-num 8 \
          --tokenizer-worker-num 8 \
          --enable-strict-thinking \
          --enable-hierarchical-cache \
          --hicache-ratio 3 \
          --reasoning-parser deepseek-v4 \
          --tool-call-parser deepseekv4 \

# decode
GLOO_SOCKET_IFNAME=eth0 \
      NCCL_SOCKET_IFNAME=eth0 \
      SGLANG_REQUEST_STATE_WAIT_TIMEOUT=600 \
      MC_TE_METRIC=1 \
      MC_TE_METRIC_INTERVAL_SECONDS=1 \
      MC_INTRA_NVLINK=true \
      MC_INTRANODE_NVLINK=true \
      CUDA_VISIBLE_DEVICES=2,3 \
      SGLANG_MOONCAKE_CUSTOM_MEM_POOL=INTRA_NODE_NVLINK \
      SGLANG_DISAGGREGATION_ALL_CP_RANKS_TRANSFER=1 \
      sglang serve \
      --trust-remote-code \
      --model-path /models/deepseek-ai/DeepSeek-V4-Flash-0731 \
      --tp 2 \
      --dp 2 \
      --ep 2 \
      --enable-dp-attention \
      --enable-dp-lm-head \
      --moe-a2a-backend megamoe \
      --enable-w4a4-mxfp4-megamoe \
      --moe-runner-backend flashinfer_mxfp4 \
      --disaggregation-mode decode \
      --disaggregation-transfer-backend mooncake \
      --speculative-algorithm DSPARK \
      --max-running-requests 512 \
      --swa-full-tokens-ratio 0.05 \
      --mem-fraction-static 0.85 \
      --detokenizer-worker-num 8 \
      --tokenizer-worker-num 8 \
      --enable-strict-thinking \
      --reasoning-parser deepseek-v4 \
      --tool-call-parser deepseekv4 \
```
image

## Precision
gsm8k
image

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #35451290254](https://github.com/sgl-project/sglang/actions/runs/35451290254)
Latest PR Test (Extra): :x: [Run #35451290181](https://github.com/sgl-project/sglang/actions/runs/35451290181)
Latest PR Test (AMD ROCm 10): :x: [Run #35451290287](https://github.com/sgl-project/sglang/actions/runs/35451290287)

### CODE REVIEW

File: python/sglang/srt/layers/attention/deepseek_v4_backend.py [hunk 1]
```diff
@@ -831,6 +831,14 @@ def init_forward_metadata_prefill(
             assert cp_metadata is not None
             padded_num_tokens = sum(cp_metadata.per_rank_actual_token)
 
+        # Eager CP writes only logical tokens. BCG gathers a fixed global
+        # bucket, including dummy rows, into the KV/compressor store paths.
+        # Keep those write buffers fixed while expanding causal metadata from
+        # the real extend lengths and partitioning queries over physical rows.
+        num_write_tokens = (
```
Comment (Oasis-Git) [inline python/sglang/srt/layers/attention/deepseek_v4_backend.py:838]: Looks like dead code

File: test/registered/cp/test_dsv4_prefill_cp_bcg.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+            DeepseekV4AttnBackend,
+        )
+
+        batch = SimpleNamespace(forward_mode=ForwardMode.EXTEND)
+        for cp_active in (False, True):
+            for forced in (False, True):
+                for num_tokens in (256, 11673, 11674, 16384, 32768):
+                    with (
+                        patch(
+                            "sglang.srt.layers.attention.deepseek_v4_backend.is_cp_v2_active",
```
Comment (Oasis-Git) [inline test/registered/cp/test_dsv4_prefill_cp_bcg.py]: Test fail

File: python/sglang/srt/layers/attention/deepseek_v4_backend.py (hunk 1 above)
Comment (BJWang-ant) [reply python/sglang/srt/layers/attention/deepseek_v4_backend.py]: > Looks like dead code

It has been fixed. This code needs to be used to distinguish the padding token and the actual token when forwarding metadata in BCG.

File: test/registered/cp/test_dsv4_prefill_cp_bcg.py (hunk 2 above)
Comment (BJWang-ant) [reply test/registered/cp/test_dsv4_prefill_cp_bcg.py]: > Test fail

fixed done.

**Review by Oasis-Git — APPROVED** [review]
Please remove unnecessary comments generated by ai

**Comment by BJWang-ant** [comment]
> Please remove unnecessary comments generated by ai

ok

**Comment by Oasis-Git** [comment]
base pass: https://github.com/sgl-project/sglang/actions/runs/35451290254
